src/bootstrap/lib.rs

```diff
@@ -455,6 +455,31 @@ impl Build {
             job::setup(self);
         }
 
+        // Duplicated with Python.
+        //
+        // This builds bootstrap *again*; this is quite quick usually as the
+        // dependencies and such have already been built. Needed because `cargo
+        // run` will not build binaries other than "main.rs", and we need those
+        // built. They also depend on bootstrap so we cannot currently make them
+        // stage0 bootstrap tools without recompiling bootstrap's dependencies.
+        let dir = &self.config.rustbuild_bin_directory;
+        let mut cmd = Command::new(&self.config.initial_cargo);
+        cmd.env("CARGO_TARGET_DIR", &dir.parent().unwrap())
+            // `CARGO_BUILD_TARGET` breaks bootstrap build.
+            // See also: <https://github.com/rust-lang/rust/issues/70208>.
+            .env_remove("CARGO_BUILD_TARGET")
+            .env("RUSTC", &self.initial_rustc)
+            .arg("build")
+            .arg("--manifest-path")
+            .stdout(std::process::Stdio::null())
+            .arg(PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("Cargo.toml"));
```

Comment:
I must admit, I don't really understand the motivation of this PR, but blocks like this seem pretty weird. This seems like it's bending over really far backwards to try to get `cargo run` to work. Cargo was never intended to be a full-fledged general purpose build system though, and building the compiler is just compilcated due to how much stuff needs to get built in various ways. This isn't the end of the world to have this here but this feels ripe for bugs and other issues. 

(and this is just the beginning of trying to get `cargo run` to work...)

Reply:
Much of this discussion has been happening on Zulip. This PR probably doesn't make sense without having followed along there.

I will say that, regardless of where Cargo started or what was intended back in the day, it needs to *become* a general build system eventually, or at least become far closer to being one. Because Cargo is how every single Rust programmer is told to interact with their Rust projects. So when Cargo can't handle a task, it's suddenly *astoundingly* painful to do that task with a Rust project. This doesn't need to happen today or tomorrow, but it needs to be a long term goal.

Reply:
> Much of this discussion has been happening on Zulip. 

https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/RIIR.20contributor.20entry.20point

Reply:
> Much of this discussion has been happening on Zulip

Ah ok, thanks for the heads up. For reviewing this though I don't have time to read up on the whole stream, so I'm gonna stick to what's in this PR.

> it needs to become a general build system eventually

This PR is probably not the best location to push this idea.